cmd/bad-key-revoker/main.go

```diff
@@ -0,0 +1,360 @@
+			initialID,
+			bkr.uncheckedBatchSize,
+		)
+		if err != nil {
+			if db.IsNoRows(err) {
+				return uncheckedRows, nil
+			}
+			return nil, err
+		}
+		uncheckedRows = append(uncheckedRows, batch...)
+		if len(batch) != bkr.uncheckedBatchSize {
+			return uncheckedRows, nil
+		}
+		initialID = batch[len(batch)-1].ID
+	}
+}
+
+type unrevoked struct {
+	Serial         string
+	DER            []byte
+	RegistrationID int64
+	RevokedBy      int64
+}
+
+var revokerName = "bad-key-revoker"
+
+func (bkr *badKeyRevoker) findUnrevoked(row unchecked) ([]unrevoked, error) {
+	var unrevokedCerts []unrevoked
+	initialID := 0
+	for {
+		var batch []struct {
+			ID         int
+			CertSerial string
+		}
+		_, err := bkr.dbMap.Select(
+			&batch,
+			"SELECT id, certserial FROM keyHashToSerial WHERE keyHash = ? and id > ? ORDER BY id LIMIT ?",
+			row.KeyHash,
+			initialID,
+			bkr.certificatesBatchSize,
+		)
+		if err != nil {
+			if db.IsNoRows(err) {
+				return unrevokedCerts, nil
+			}
+			return nil, err
+		}
+		for _, serial := range batch {
+			var unrevokedCert unrevoked
+			err = bkr.dbMap.SelectOne(
+				&unrevokedCert,
+				`SELECT cs.serial, c.registrationID, c.der FROM certificateStatus AS cs
+				JOIN certificates AS c
+				ON cs.serial = c.serial
+				WHERE cs.serial = ? AND cs.isExpired = false AND cs.status != ?`,
+				serial.CertSerial,
+				string(core.StatusRevoked),
+			)
+			if err != nil {
+				if db.IsNoRows(err) {
+					continue
+				}
+				return nil, err
+			}
+			unrevokedCert.RevokedBy = row.RevokedBy
+			unrevokedCerts = append(unrevokedCerts, unrevokedCert)
+		}
+		if len(batch) != bkr.certificatesBatchSize {
+			return unrevokedCerts, nil
+		}
+		initialID = batch[len(batch)-1].ID
+	}
+}
+
+func (bkr *badKeyRevoker) markRowChecked(row unchecked) error {
+	_, err := bkr.dbMap.Exec("UPDATE blockedKeys SET extantCertificatesChecked = true WHERE keyHash = ?", row.KeyHash)
+	return err
+}
+
+func (bkr *badKeyRevoker) resolveContacts(ids []int64) (map[int64]string, error) {
+	idToEmail := map[int64]string{}
+	for _, id := range ids {
+		var emails struct {
+			Contact []string
+		}
+		err := bkr.dbMap.SelectOne(&emails, "SELECT contact FROM registrations WHERE id = ?", id)
+		if err != nil {
+			return nil, err
+		}
+		if len(emails.Contact) == 0 {
+			continue
+		}
+		idToEmail[id] = strings.TrimPrefix(emails.Contact[0], "mailto:")
+	}
+	return idToEmail, nil
+}
+
+var emailTemplate = template.Must(template.New("email-template").Parse(`Hello,
+
+The public key associated with certificates which you have issued has been marked as compromised. As such we are required to revoke any certificates which contain this public key.
+
+The following currently unexpired certificates that you've issued contain this public key and have been revoked:
+{{range . -}}
+{{.}}
+{{end}}`))
+
+var emailSubject = "Certificates you've issued have been revoked due to key compromise"
+
+var maxSerials = 100
```

Comment:
```suggestion
const maxSerials = 100
```

Also, this can move inside the function that uses it.

Reply:
This is used in testing.